enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,478 @@
+6. The installer bootstraps the cluster.
+7. Bootstrapping runs the machine-config-operator (MCO) `render`
+   command, which generates MachineConfig manifests with low
+   precedence.
+8. Bootstrapping uploads both sets of MachineConfig manifests, one
+   after the other.
+9. The machine-config-operator (MCO) applies MachineConfigs to nodes
+   in the cluster.
+10. Bootstrapping finishes and the cluster is launched.
+11. If the MCO does not apply KubeletConfig in step 9, it must do it
+    here.
+12. The cluster has complete partitioning configuration for management
+    workloads.
+13. Kubelet starts and sees the config file enabling partitioning
+    (delivered in the MachineConfig manifest generated by
+    PAO). Kubelet advertises the workload resource type on the Node
+    and mutates static pods with partitioning annotations.
+14. The admission plugin uses the workload types on the new enablement
+    API to decide when to mutate pods.
+15. CRI-O sees pods with workload annotations and uses the resource
+    request to set cpushares and cpuset.
+16. On day 2 PAO, is installed and takes ownership of the
+    PerformanceProfile CRs.
+
+#### Day 2: Modify Reserved cpuset for a PerformanceProfile
+
+In a cluster with partitioning enabled and fully configured, modifying
+the reserved cpuset for a PerformanceProfile is safe because the node
+is rebooted when the new MachineConfig is applied.
+
+1. The user updates the cpuset in the PerformanceProfile(s).
+2. PAO generates new MachineConfigs, adding the cpuset information
+   from the PerformanceProfiles to the other partitioning info from
+   the new enablement API.
+3. The MCO rolls out new MachineConfigs, rebooting nodes as it goes.
+4. The cluster has complete partitioning configuration for management
+   workloads and the management workloads are partitioned due to the
+   nodes rebooting.
+5. Kubelet starts and sees the config file enabling partitioning. It
+   advertises the workload resource type on the Node and mutates
+   static pods with partitioning annotations.
+6. Admission plugin uses the workload types on the workloads CR to
+   decide when to mutate pods.
+7. CRI-O sees pods with workload annotations and uses the resource
+   request to set cpushares and cpuset.
+
+#### Day 2: Add a New Node to an Existing MachineConfigPool
+
+In a cluster with partitioning enabled and fully configured, adding a
+new node to an existing MachineConfigPool is safe because the
+MachineConfig will set up kubelet and CRI-O on the host correctly.
+
+1. The MCO applies the MachineConfigs to the node.
+2. Kubelet starts and sees the config file enabling partitioning. It
+   advertises the workload resource type on the Node and mutates
+   static pods with partitioning annotations.
+3. CRI-O sees pods with workload annotations and uses the resource
+   request to set cpushares and cpuset.
+
+#### Day 2: Add a New MachineConfigPool
+
+Adding a new MachineConfigPool is safe only if the MachineConfigPool
```

Comment:
all custom MCPs inherit from worker, so they must have worker-MCP based configs.

This is somewhat of an issue in the kubeletconfig use case today. The kubeletconfig doesn't really order based on pool priority but implied MCO naming-schema priority, which means if a worker pool and a custom pool both have custom kubeletconfigs, the one that actually gets laid down is alphabetically ordered, meaning that `custom, worker` will result in worker configs, and `worker,zz-custom` will result in zz-custom configs.

The suggestion today I have is not to have base worker kubeletconfigs if you have custom pools that need different ones. All nodes that require separate kubeletconfigs must be in a custom pool (empty worker pool). I think we should reconsider this section

Reply:
That's not any different from what we have today, right? I don't know what the best practices for using PerformanceProfiles is, but I don't expect that to change dramatically as a result of this work.

@browsell or @MarSik may have more insight into how they're being used today.

Reply:
No, that is the behaviour today. I just want to make sure we are not overwriting our own configs. Custom pools need to be handled very carefully

Reply:
Yeah, this is not going to change.

Btw the fact that KubeletConfigs cannot be layered is already causing us some headaches. It is a different feature altogether, but we need to set cpu manager stuff and customers want to enable unsafe syscalls. And there just is no way to do that and keep ownership of just half of the config.